### ISSUE TITLE
nosedep dependencies not applied when running a directory of tests

### ISSUE THREAD

**Issue by hohene** [Zitrax/nose-dep#9]
Looking at one of the older issues, it appears the dependencies should work when running nosetests via selecting a starting directory as opposed to running the file directly.  That seems to not be working.  We use groups and 'tests' as starting points for running suites with nosetests.
Have you gotten this to work?
Here's the test:
```
import unittest
from nosedep import depends

# set path to common libraries

class display_node_list(unittest.TestCase):
    @depends(after='test_02')
    def test_01(self):
        print "\nTest 1"

    @depends(after='test_03')
    def test_02(self):
        print "\ntest 2"

    @depends(after='test_04')
    def test_03(self):
        print "\ntest 3"

    def test_04(self):
        # This test displays a list of the nodes, type and name
        print "\ntest 4"

```
Running the file dep_test.py directly:
```
(eh-ub16) hohene@lab17:~/sandbox-github/rackhd/eh-cit-nosedep/test/tests$ nosetests --exe -s -v  --with-nosedep api-cit/eh/dep_test.py
test_04 (eh.dep_test.display_node_list) ... 
test 4
ok
test_03 (eh.dep_test.display_node_list) ... 
test 3
ok
test_02 (eh.dep_test.display_node_list) ... 
test 2
ok
test_01 (eh.dep_test.display_node_list) ... 
Test 1
ok

----------------------------------------------------------------------
Ran 4 tests in 0.002s

OK
```
Running by the directory:
```
(eh-ub16) hohene@lab17:~/sandbox-github/rackhd/eh-cit-nosedep/test/tests$ nosetests --exe -s -v  --with-nosedep api-cit/eh
test_01 (eh.dep_test.display_node_list) ... ERROR
test_02 (eh.dep_test.display_node_list) ... ERROR
test_03 (eh.dep_test.display_node_list) ... ERROR
test_04 (eh.dep_test.display_node_list) ... 
test 4
ok

======================================================================
ERROR: test_01 (eh.dep_test.display_node_list)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/emc/hohene/sandbox-github/rackhd/eh-cit-nosedep/test/env/eh-ub16/local/lib/python2.7/site-packages/nosedep.py", line 323, in error_test
    raise Exception(res)
Exception: Required test 'test_02' did not run (does it exist?)

======================================================================
ERROR: test_02 (eh.dep_test.display_node_list)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/emc/hohene/sandbox-github/rackhd/eh-cit-nosedep/test/env/eh-ub16/local/lib/python2.7/site-packages/nosedep.py", line 323, in error_test
    raise Exception(res)
Exception: Required test 'test_03' did not run (does it exist?)

======================================================================
ERROR: test_03 (eh.dep_test.display_node_list)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/emc/hohene/sandbox-github/rackhd/eh-cit-nosedep/test/env/eh-ub16/local/lib/python2.7/site-packages/nosedep.py", line 323, in error_test
    raise Exception(res)
Exception: Required test 'test_04' did not run (does it exist?)

----------------------------------------------------------------------
Ran 4 tests in 0.004s

FAILED (errors=3)
```

**Comment by Zitrax** [comment]
Hi, been on vacation so just read this. I'll get back to you when I have had a closer look - just letting you know I have seen your report.

**Comment by larry-dean** [comment]
I work with @hohene and she ask me to include some of the work I did to debug the issue. In the case of starting the tests from the parent directory, it appears that there is an additional  "ContextSuite' level in the test object presented to prepareTest(). 

    --t- <nose.suite.LazySuite tests=generator (139873539047120)>
    -tt- <nose.suite.ContextSuite context=v2_0>
    --t- <nose.suite.ContextSuite context=v2_0>
    -tt- <nose.suite.ContextSuite context=v2_0.dep_test>
    -ttt <nose.suite.ContextSuite context=display_rackhd_node_list>

Below is a 'hack' I put together to get around the issue and some traces of the results. I hope it helps.

Hacked code

    def prepareTest(self, test):
        """Prepare and determine test ordering"""
        print "********* Entering prepareTest "
        time.sleep(1)
        all_tests = {}
        for t in test:
            print "--t- {}".format(t)
            for tt in t:
                print "-tt- {}".format(tt)
                if tt.context is None:
                    print "*** continue - {}".format(tt)
                    continue
                try:
                    if isinstance(tt, ContextSuite):  # MethodTestCase
                        for ttt in tt:
                            print "-ttt {}".format(ttt)
                            if ttt.context and not isinstance(ttt, ContextSuite):
                                print "********* MethodTestCase(tt) - {}".format(tt)
                                all_tests[tt.context.__name__] = self.prepare_suite(tt)
                                setattr(all_tests[tt.context.__name__], 'nosedep_run', True)
                            else:
                                print "********* MethodTestCase (ttt) - {}".format(ttt)
                                all_tests[ttt.context.__name__] = self.prepare_suite(ttt)
                                setattr(all_tests[ttt.context.__name__], 'nosedep_run', True)
                             
                    else:  # FunctionTestCase
                        print "********* FunctionTestCase - {}".format(tt)
                        all_tests[tt.test.test.__name__] = tt
                except AttributeError:
                    print "*************** except"
                    # This exception is confusing - reraise the original one
                    reraise(tt.test.exc_class, tt.test.exc_val, tt.test.tb)
        print "********* exiting prepareTest  and call to orderTests "
        return self.orderTests(all_tests, test)


Trace of running test directly


    python run_tests.py -stack 10 -test tests/api-cit/v2_0/dep_test.py


    ********* Entering prepareTest
    nose.suite: DEBUG: precache is [<nose.suite.ContextSuite context=v2_0.dep_test>]
    --t- <nose.suite.ContextSuite context=v2_0.dep_test>
    nose.suite: DEBUG: precache is [<nose.suite.ContextSuite context=display_rackhd_node_list>]
    -tt- <nose.suite.ContextSuite context=display_rackhd_node_list>
    nose.suite: DEBUG: precache is [Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_01>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_02>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_03>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_04>)]
    -ttt test_01 (v2_0.dep_test.display_rackhd_node_list)
    ********* MethodTestCase(tt) - <nose.suite.ContextSuite context=display_rackhd_node_list>
    ********* Entering prepare_suite
    nose.suite: DEBUG: precache is [Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_01>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_02>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_03>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_04>)]
    ********* prepare_suite Before orderTests call
    nose.suite: DEBUG: precache is [Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_01>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_02>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_03>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_04>)]
    >>>   test_01 (v2_0.dep_test.display_rackhd_node_list)
    >>>   test_02 (v2_0.dep_test.display_rackhd_node_list)
    >>>   test_03 (v2_0.dep_test.display_rackhd_node_list)
    >>>   test_04 (v2_0.dep_test.display_rackhd_node_list)
    ********* Entering orderTests
    ********* orderTests complete
    >>>   test_04
    >>>   test_01
    >>>   test_02
    >>>   test_03
    -ttt test_02 (v2_0.dep_test.display_rackhd_node_list)
    ********* MethodTestCase(tt) - <nose.suite.ContextSuite context=display_rackhd_node_list>
    ********* Entering prepare_suite
    nose.suite: DEBUG: precache is [Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_04>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_03>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_02>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_01>)]
    ********* prepare_suite Before orderTests call
    nose.suite: DEBUG: precache is [Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_04>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_03>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_02>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_01>)]
    >>>   test_04 (v2_0.dep_test.display_rackhd_node_list)
    >>>   test_03 (v2_0.dep_test.display_rackhd_node_list)
    >>>   test_02 (v2_0.dep_test.display_rackhd_node_list)
    >>>   test_01 (v2_0.dep_test.display_rackhd_node_list)
    ********* Entering orderTests
    ********* orderTests complete
    >>>   test_04
    >>>   test_01
    >>>   test_02
    >>>   test_03
    -ttt test_03 (v2_0.dep_test.display_rackhd_node_list)
    ********* MethodTestCase(tt) - <nose.suite.ContextSuite context=display_rackhd_node_list>
    ********* Entering prepare_suite
    nose.suite: DEBUG: precache is [Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_04>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_03>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_02>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_01>)]
    ********* prepare_suite Before orderTests call
    nose.suite: DEBUG: precache is [Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_04>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_03>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_02>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_01>)]
    >>>   test_04 (v2_0.dep_test.display_rackhd_node_list)
    >>>   test_03 (v2_0.dep_test.display_rackhd_node_list)
    >>>   test_02 (v2_0.dep_test.display_rackhd_node_list)
    >>>   test_01 (v2_0.dep_test.display_rackhd_node_list)
    ********* Entering orderTests
    ********* orderTests complete
    >>>   test_04
    >>>   test_01
    >>>   test_02
    >>>   test_03
    -ttt test_04 (v2_0.dep_test.display_rackhd_node_list)
    ********* MethodTestCase(tt) - <nose.suite.ContextSuite context=display_rackhd_node_list>
    ********* Entering prepare_suite
    nose.suite: DEBUG: precache is [Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_04>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_03>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_02>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_01>)]
    ********* prepare_suite Before orderTests call
    nose.suite: DEBUG: precache is [Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_04>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_03>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_02>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_01>)]
    >>>   test_04 (v2_0.dep_test.display_rackhd_node_list)
    >>>   test_03 (v2_0.dep_test.display_rackhd_node_list)
    >>>   test_02 (v2_0.dep_test.display_rackhd_node_list)
    >>>   test_01 (v2_0.dep_test.display_rackhd_node_list)
    ********* Entering orderTests
    ********* orderTests complete
    >>>   test_04
    >>>   test_01
    >>>   test_02
    >>>   test_03
    ********* exiting prepareTest  and call to orderTests
    ********* Entering orderTests
    ********* orderTests complete
    >>>   display_rackhd_node_list

Trace of running test from parent directory


    python run_tests.py -stack 10 -test tests/api-cit/v2_0


    ********* Entering prepareTest
    nose.suite: DEBUG: precache is [<nose.suite.LazySuite tests=generator (139873539047120)>, <nose.suite.ContextSuite context=v2_0>]
    --t- <nose.suite.LazySuite tests=generator (139873539047120)>
    nose.suite: DEBUG: precache is []
    nose.loader: DEBUG: load from dir /emc/deanl1/src2/hohene/test/tests/api-cit
    nose.importer: DEBUG: Add path /emc/deanl1/src2/hohene/test/tests/api-cit
    nose.importer: DEBUG: Add path /emc/deanl1/src2/hohene/test/tests
    nose.importer: DEBUG: Add path /emc/deanl1/src2/hohene/test
    nose.importer: DEBUG: Add path /emc/deanl1/src2/hohene
    nose.selector: DEBUG: __init__.py matches ignoreFiles pattern; skipped
    nose.selector: DEBUG: wantDirectory /emc/deanl1/src2/hohene/test/tests/api-cit/v2_0? True
    nose.loader: DEBUG: load from /emc/deanl1/src2/hohene/test/tests/api-cit/v2_0 (None)
    nose.selector: DEBUG: Test name /emc/deanl1/src2/hohene/test/tests/api-cit/v2_0 resolved to file /emc/deanl1/src2/hohene/test/tests/api-cit/v2_0, module None, call None
    nose.selector: DEBUG: Final resolution of test name /emc/deanl1/src2/hohene/test/tests/api-cit/v2_0: file /emc/deanl1/src2/hohene/test/tests/api-cit/v2_0 module v2_0 call None
    nose.importer: DEBUG: Import v2_0 from /emc/deanl1/src2/hohene/test/tests/api-cit
    nose.importer: DEBUG: Add path /emc/deanl1/src2/hohene/test/tests/api-cit
    nose.importer: DEBUG: Add path /emc/deanl1/src2/hohene/test/tests
    nose.importer: DEBUG: Add path /emc/deanl1/src2/hohene/test
    nose.importer: DEBUG: Add path /emc/deanl1/src2/hohene
    nose.importer: DEBUG: find module part v2_0 (v2_0) in ['/emc/deanl1/src2/hohene/test/tests/api-cit']
    nose.importer: DEBUG: sys.modules has v2_0 as <module 'v2_0' from '/emc/deanl1/src2/hohene/test/tests/api-cit/v2_0/__init__.pyc'>
    nose.importer: DEBUG: module already loaded? mod: /emc/deanl1/src2/hohene/test/tests/api-cit/v2_0 new: /emc/deanl1/src2/hohene/test/tests/api-cit/v2_0
    nose.loader: DEBUG: Load from module <module 'v2_0' from '/emc/deanl1/src2/hohene/test/tests/api-cit/v2_0/__init__.pyc'>
    nose.selector: DEBUG: wantModule <module 'v2_0' from '/emc/deanl1/src2/hohene/test/tests/api-cit/v2_0/__init__.pyc'>? False
    nose.loader: DEBUG: Load tests from module path /emc/deanl1/src2/hohene/test/tests/api-cit/v2_0?
    nose.loader: DEBUG: path: /emc/deanl1/src2/hohene/test/tests/api-cit/v2_0 os.path.realpath(/emc/deanl1/src2/hohene/test/tests/api-cit/v2_0): /emc/deanl1/src2/hohene/test/tests/api-cit/v2_0
    nose.loader: DEBUG: load from dir /emc/deanl1/src2/hohene/test/tests/api-cit/v2_0
    nose.importer: DEBUG: Add path /emc/deanl1/src2/hohene/test/tests/api-cit/v2_0
    nose.importer: DEBUG: Add path /emc/deanl1/src2/hohene/test/tests/api-cit
    nose.importer: DEBUG: Add path /emc/deanl1/src2/hohene/test/tests
    nose.importer: DEBUG: Add path /emc/deanl1/src2/hohene/test
    nose.importer: DEBUG: Add path /emc/deanl1/src2/hohene
    nose.selector: DEBUG: __init__.py matches ignoreFiles pattern; skipped
    nose.selector: DEBUG: __init__.pyc matches ignoreFiles pattern; skipped
    nose.selector: DEBUG: wantFile /emc/deanl1/src2/hohene/test/tests/api-cit/v2_0/nodes_eh.py? None
    nose.selector: DEBUG: wantFile /emc/deanl1/src2/hohene/test/tests/api-cit/v2_0/nodes_eh.pyc? False
    nose.selector: DEBUG: wantFile /emc/deanl1/src2/hohene/test/tests/api-cit/v2_0/dep_test.py? True
    nose.selector: DEBUG: wantFile /emc/deanl1/src2/hohene/test/tests/api-cit/v2_0/dep_test.pyc? False
    nose.selector: DEBUG: wantFile /emc/deanl1/src2/hohene/test/tests/api-cit/v2_0/nodes_tests_fit.py.old? False
    nose.suite: DEBUG: Create suite for <nose.suite.ContextList object at 0x7f36d7b6c5d0>
    nose.suite: DEBUG: tests <nose.suite.ContextList object at 0x7f36d7b6c5d0> context <module 'v2_0' from '/emc/deanl1/src2/hohene/test/tests/api-cit/v2_0/__init__.pyc'>
    nose.suite: DEBUG: Context suite for <nose.suite.ContextList object at 0x7f36d7b6c5d0> (<module 'v2_0' from '/emc/deanl1/src2/hohene/test/tests/api-cit/v2_0/__init__.pyc'>) (139873524041360)
    nose.suite: DEBUG: suite <nose.suite.ContextSuite context=v2_0> has context v2_0
    nose.suite: DEBUG: get ancestry <module 'v2_0' from '/emc/deanl1/src2/hohene/test/tests/api-cit/v2_0/__init__.pyc'>
    -tt- <nose.suite.ContextSuite context=v2_0>
    nose.suite: DEBUG: precache is []
    --t- <nose.suite.ContextSuite context=v2_0>
    nose.suite: DEBUG: precache is [<nose.suite.ContextSuite context=v2_0.dep_test>]
    -tt- <nose.suite.ContextSuite context=v2_0.dep_test>
    nose.suite: DEBUG: precache is [<nose.suite.ContextSuite context=display_rackhd_node_list>]
    -ttt <nose.suite.ContextSuite context=display_rackhd_node_list>
    ********* MethodTestCase (ttt) - <nose.suite.ContextSuite context=display_rackhd_node_list>
    ********* Entering prepare_suite
    nose.suite: DEBUG: precache is [Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_01>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_02>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_03>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_04>)]
    ********* prepare_suite Before orderTests call
    nose.suite: DEBUG: precache is [Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_01>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_02>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_03>), Test(<v2_0.dep_test.display_rackhd_node_list testMethod=test_04>)]
    >>>   test_01 (v2_0.dep_test.display_rackhd_node_list)
    >>>   test_02 (v2_0.dep_test.display_rackhd_node_list)
    >>>   test_03 (v2_0.dep_test.display_rackhd_node_list)
    >>>   test_04 (v2_0.dep_test.display_rackhd_node_list)
    ********* Entering orderTests
    ********* orderTests complete
    >>>   test_04
    >>>   test_01
    >>>   test_02
    >>>   test_03
    ********* exiting prepareTest  and call to orderTests
    ********* Entering orderTests
    ********* orderTests complete
    >>>   display_rackhd_node_list

**Comment by Zitrax** [comment]
Thanks for the analysis, it helped.

To be honest I do not have the full overview of the nose internals in this case, but as you mentioned there is an extra level - so in my commit I just go down one level when we have a None context. I added an assert to verify the assumption that we have only one item on that level. This is at least true for all the current unit tests (and testing directory tests are now part of the unit tests).

Could you try out my patch and see if it helps for you.

( However I see I need to modify it, since it fails on Python 3 )

**Comment by Zitrax** [comment]
Python 3 issue fixed.

**Comment by larry-dean** [comment]
The correction does not seem to have any effect. I will analyze the issue on Monday.

**Comment by larry-dean** [comment]
Sorry for the delay. I ran your changes against the following cases with the same results as before the change. 

Passing Case  (tests/api-cit/v2_0/dep_test.py)

    **** OBJECT DUMP ****
       ***** <nose.suite.ContextSuite context=v2_0.dep_test>
         ***** <nose.suite.ContextSuite context=v2_0.dep_test>
           ***** <nose.suite.ContextSuite context=display_rackhd_node_list>
             ***** test_01 (v2_0.dep_test.display_rackhd_node_list)
             ***** test_02 (v2_0.dep_test.display_rackhd_node_list)
             ***** test_03 (v2_0.dep_test.display_rackhd_node_list)
             ***** test_04 (v2_0.dep_test.display_rackhd_node_list)

 Failing Case (tests/api-cit/v2_0)
 
    **** OBJECT DUMP ****
    ***** <nose.suite.ContextSuite context=v2_0>
     ***** <nose.suite.ContextSuite context=v2_0>
       ***** <nose.suite.ContextSuite context=v2_0.dep_test>
         ***** <nose.suite.ContextSuite context=display_rackhd_node_list>
           ***** test_01 (v2_0.dep_test.display_rackhd_node_list)
           ***** test_02 (v2_0.dep_test.display_rackhd_node_list)
           ***** test_03 (v2_0.dep_test.display_rackhd_node_list)
           ***** test_04 (v2_0.dep_test.display_rackhd_node_list)

Another failing case (tests/api-cit)

    **** OBJECT DUMP ****
    ***** <nose.suite.ContextSuite context=v2_0>
     ***** <nose.suite.LazySuite tests=generator (139822303550224)>
       ***** <nose.suite.ContextSuite context=v2_0>
     ***** <nose.suite.ContextSuite context=v2_0>
       ***** <nose.suite.ContextSuite context=v2_0.dep_test>
         ***** <nose.suite.ContextSuite context=display_rackhd_node_list>
           ***** test_01 (v2_0.dep_test.display_rackhd_node_list)
           ***** test_02 (v2_0.dep_test.display_rackhd_node_list)
           ***** test_03 (v2_0.dep_test.display_rackhd_node_list)
           ***** test_04 (v2_0.dep_test.display_rackhd_node_list)

As you can see there is at least one additional level presented in the failing cases. 


I wrote a re-enterent function that handles all the cases above. It runs our tests as well as all your tests individually (I have environmental issues with tests.py). It also processes multiple test files (the example is after sorting is complete).

    **** OBJECT DUMP ****
       ***** <nose.suite.ContextSuite context=test.tests.nosedep.test_scripts>
         ***** <nose.suite.ContextSuite context=TestNoseDecoratedMethod>
           ***** test.tests.nosedep.test_scripts.decorated_method_tests.TestNoseDecoratedMethod.test_cd_a
           ***** test.tests.nosedep.test_scripts.decorated_method_tests.TestNoseDecoratedMethod.test_cd_c
           ***** test.tests.nosedep.test_scripts.decorated_method_tests.TestNoseDecoratedMethod.test_cd_b
         ***** <nose.suite.ContextSuite context=TestNoseDecoratedTestCaseChild>
           ***** test_cdt_a (test.tests.nosedep.test_scripts.decorated_method_subclass_testcase.TestNoseDecoratedTestCaseChild)
           ***** test_cdt_c (test.tests.nosedep.test_scripts.decorated_method_subclass_testcase.TestNoseDecoratedTestCaseChild)
           ***** test_cdt_b (test.tests.nosedep.test_scripts.decorated_method_subclass_testcase.TestNoseDecoratedTestCaseChild)
         ***** <nose.suite.ContextSuite context=TestNoseUndecoratedMethod>
           ***** test.tests.nosedep.test_scripts.undecorated_method_tests.TestNoseUndecoratedMethod.test_c_a
           ***** test.tests.nosedep.test_scripts.undecorated_method_tests.TestNoseUndecoratedMethod.test_c_b
         ***** <nose.suite.ContextSuite context=test.tests.nosedep.test_scripts.decorated_functional_tests>
           ***** test.tests.nosedep.test_scripts.decorated_functional_tests.test_dft_f
           ***** test.tests.nosedep.test_scripts.decorated_functional_tests.test_dft_b
           ***** test.tests.nosedep.test_scripts.decorated_functional_tests.test_dft_e
           ***** test.tests.nosedep.test_scripts.decorated_functional_tests.test_dft_a
           ***** test.tests.nosedep.test_scripts.decorated_functional_tests.test_dft_c
           ***** test.tests.nosedep.test_scripts.decorated_functional_tests.test_dft_d
         ***** <nose.suite.ContextSuite context=test.tests.nosedep.test_scripts.undecorated_functional_tests>
           ***** test.tests.nosedep.test_scripts.undecorated_functional_tests.test_x
           ***** test.tests.nosedep.test_scripts.undecorated_functional_tests.test_y

           
The code follows. It is highly instrumented and contains a little dump utility (a_test_thingy()).

    def a_print_thingy(self, obj, cnt = 0):
        if cnt == 0:
            print "**** OBJECT DUMP ****"
            cnt = 2
        print '{} ***** {}'.format(' ' * cnt, obj)
        for t in obj:
            try:
                self.a_print_thingy(t, cnt + 2)
            except TypeError:
               pass
        return

    def prepareTest(self, test):
        """Prepare and determine test ordering"""
        all_tests = {}
        self.a_print_thingy(test)

        # When passing a directory to nose we have an extra
        # top level that we need to enter.
        print '>>>> test: {}'.format(test)
        all_tests = self.find_next_level(test, all_tests, 1)

        #return self.orderTests(all_tests, test)
        hello =  self.orderTests(all_tests, test)
        self.a_print_thingy(hello)
        return hello

    def find_next_level(self, test, all_tests, cnt):
        s = ' '
        print '{} >>>> {}'.format(s * cnt, test)
        for t in test:
            try:
                self.find_next_level(t, all_tests, cnt + 2)
            except TypeError as err:
                print '{} ^^^^^^^ {}'.format(s * cnt, err)
                if isinstance(test, ContextSuite):  # MethodTestCase
                    print '{} >>>> MethodTestCase'.format(s * cnt)
                    all_tests[test.context.__name__] = self.prepare_suite(test)
                    setattr(all_tests[test.context.__name__], 'nosedep_run', True)
                else:  # FunctionTestCase
                    print ' {} >>>> FunctionTestCase'.format(s * cnt)
                    all_tests[test.test.test.__name__] = test
                break
            except AttributeError:
                # This exception is confusing - reraise the original one
                reraise(test.test.exc_class, test.test.exc_val, test.test.tb)
        return all_tests
 
An example of the output for the (tests/api-cit) case:

    **** OBJECT DUMP ****
       ***** <nose.suite.ContextSuite context=v2_0>
         ***** <nose.suite.LazySuite tests=generator (139822943590992)>
           ***** <nose.suite.ContextSuite context=v2_0>
         ***** <nose.suite.ContextSuite context=v2_0>
           ***** <nose.suite.ContextSuite context=v2_0.dep_test>
             ***** <nose.suite.ContextSuite context=display_rackhd_node_list>
               ***** test_01 (v2_0.dep_test.display_rackhd_node_list)
               ***** test_02 (v2_0.dep_test.display_rackhd_node_list)
               ***** test_03 (v2_0.dep_test.display_rackhd_node_list)
               ***** test_04 (v2_0.dep_test.display_rackhd_node_list)
    >>>> test: <nose.suite.ContextSuite context=v2_0>
      >>>> <nose.suite.ContextSuite context=v2_0>
        >>>> <nose.suite.LazySuite tests=generator (139822943590992)>
        >>>> <nose.suite.ContextSuite context=v2_0>
          >>>> <nose.suite.ContextSuite context=v2_0.dep_test>
            >>>> <nose.suite.ContextSuite context=display_rackhd_node_list>
              >>>> test_01 (v2_0.dep_test.display_rackhd_node_list)
            ^^^^^^^ 'Test' object is not iterable
            >>>> MethodTestCase
    **** OBJECT DUMP ****
       ***** <nose.suite.ContextSuite context=v2_0>
         ***** <nose.suite.ContextSuite context=display_rackhd_node_list>
           ***** test_04 (v2_0.dep_test.display_rackhd_node_list)
           ***** test_03 (v2_0.dep_test.display_rackhd_node_list)
           ***** test_02 (v2_0.dep_test.display_rackhd_node_list)
           ***** test_01 (v2_0.dep_test.display_rackhd_node_list)

**Comment by larry-dean** [comment]
@Zitrax,  would you prefer i create a and branch and a PR?

**Comment by larry-dean** [comment]
(lab2)deanl1@lab02:~/src2/nosedep-bug$ python tests.py
    ............................
    ----------------------------------------------------------------------
    Ran 28 tests in 0.814s

**Comment by Zitrax** [comment]
Sounds great.

Would be nice to have a unit test that covers the extra case when my change did not help but yours did (since my added directory unit case passed with my change).

**Comment by larry-dean** [comment]
I created a test case that reproduces the failure. I will create a PR with all the changes.

**Comment by Zitrax** [comment]
I have now included this change in release v0.6.
